Oracle 18.4.0-XE: setPassword.sh fails for PDBADMIN

**Describe the issue**
Setting a new password for SYS and SYSTEM works, but fails for PDBADMIN:
[oracle@localhost ~]$ /home/oracle/setPassword.sh Not_Very_N1ce_Pass
The Oracle base remains unchanged with value /opt/oracle

SQL*Plus: Release 18.0.0.0.0 - Production on Fri Sep 4 17:52:00 2020
Version 18.4.0.0.0

Copyright (c) 1982, 2018, Oracle.  All rights reserved.


Connected to:
Oracle Database 18c Express Edition Release 18.0.0.0.0 - Production
Version 18.4.0.0.0

SQL>
User altered.

SQL>
User altered.

SQL>   2  XEPDB1
*
ERROR at line 2:
ORA-02248: invalid option for ALTER SESSION


SQL>       ALTER USER PDBADMIN IDENTIFIED BY "Not_Very_N1ce_Pass"
                 *
ERROR at line 1:
ORA-01918: user 'PDBADMIN' does not exist


SQL> Disconnected from Oracle Database 18c Express Edition Release 18.0.0.0.0 - Production` 

**Environment (please complete the following information):**

- Host OS: Windows 10
- Vagrant version: 2.2.10
- VirtualBox version: 6.1.12r139181
- Vagrant project:  OracleDatabase/18.4.0-XE

**Additional information**

Looks like there is an extra pdb with a long name, causing the error when trying to find the PDB name:
[oracle@localhost ~]$ ls -dl $ORACLE_BASE/oradata/$ORACLE_SID/*/ | grep -v pdbseed | awk '{print $9}' | cut -d/ -f6
AE7E20660A3D5A64E055000000000001
XEPDB1

@goranpaues-tradedoubler, @gvenzl I think I have a generic solution that will work.

To recap: The `setPassword.sh` script for the single-instance database projects (except for 11g XE) assumes that there is only a single PDB. It parses the directory names under `$ORACLE_BASE/oradata/$ORACLE_SID` to find the name of the PDB. This doesn't work correctly if one or more PDBs have been added.

Using `catcon` allows `setPassword.sh` to work correctly with either a single PDB or multiple PDBs. The script could look like:

```bash
#!/bin/bash
# LICENSE UPL 1.0
#
# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
#
# Since: November, 2016
# Author: gerald.venzl@oracle.com
# Description: Sets the password for sys, system and pdbadmin
#
# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
#

ORACLE_PWD=$1

sqlplus / as sysdba << EOF
      ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
      ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
      exit;
EOF

echo 'Setting PDBADMIN password in PDB(s) using catcon'
log_dir=$(mktemp -d)

# shellcheck disable=SC2016
"$ORACLE_HOME"/perl/bin/perl "$ORACLE_HOME"/rdbms/admin/catcon.pl \
  -l "$log_dir" -C 'CDB$ROOT PDB$SEED' -f -b setPassword \
  -- --x"ALTER USER PDBADMIN IDENTIFIED BY \"$ORACLE_PWD\""

rm -rf "$log_dir"
echo 'Done setting PDBADMIN password in PDB(s)'
```

This sets the SYS and SYSTEM passwords, and sets the PDBADMIN password in all open PDBs where the PDBADMIN user exists. (It also removes the call to `oraenv`. This isn't needed, because the environment variables are already set in the oracle user's `.bashrc`.) I've tested this for the 12.1.0.2, 12.2.0.1, 18.3.0, 18.4.0-XE, and 19.3.0 projects. 

@gvenzl, if you agree that this is a legitimate issue, and that this approach is reasonable, I'd be happy to submit a PR.

--- a/OracleDatabase/12.1.0.2/scripts/setPassword.sh
+++ b/OracleDatabase/12.1.0.2/scripts/setPassword.sh
@@ -1,26 +1,54 @@
 #!/bin/bash
-# LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2016 Oracle and/or its affiliates. All rights reserved.
-# 
+# Copyright (c) 2023 Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at
+# https://oss.oracle.com/licenses/upl.
+#
 # Since: November, 2016
 # Author: gerald.venzl@oracle.com
-# Description: Sets the password for sys, system and pdb_admin
+# Description: Sets the password for sys, system and pdbadmin
 #
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
-# 
+#
+
+# Abort on any error
+set -Eeuo pipefail
 
 ORACLE_PWD=$1
-ORACLE_SID="`grep $ORACLE_HOME /etc/oratab | cut -d: -f1`"
-ORACLE_PDB="`ls -dl $ORACLE_BASE/oradata/$ORACLE_SID/*/ | grep -v pdbseed | awk '{print $9}' | cut -d/ -f6`"
-ORAENV_ASK=NO
-source oraenv
 
 sqlplus / as sysdba << EOF
-      ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
-      ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
-      ALTER SESSION SET CONTAINER=$ORACLE_PDB;
-      ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD";
-      exit;
+  ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
+  ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
 EOF
 
+echo 'Setting PDBADMIN password in open PDBs'
+
+set_pdbadmin_pw=$(mktemp)
+trap 'rm -f "${set_pdbadmin_pw}"' EXIT
+
+sqlplus -s / as sysdba > "${set_pdbadmin_pw}" << EOF
+  SET HEADING OFF LINESIZE 120 PAGESIZE 0
+  SELECT   'ALTER SESSION SET CONTAINER = ' || name || ';' || CHR(10)
+        || 'SET SERVEROUTPUT ON' || CHR(10)
+        || 'BEGIN' || CHR(10)
+        || '  EXECUTE IMMEDIATE ''ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD"'';' || CHR(10)
+        || '  DBMS_OUTPUT.PUT_LINE (''Set PDBADMIN password in ' || name || ' PDB'');' || CHR(10)
+        || 'EXCEPTION' || CHR(10)
+        || '  WHEN OTHERS THEN' || CHR(10)
+        || '    IF SQLCODE = -1918 THEN' || CHR(10)
+        || '      DBMS_OUTPUT.PUT_LINE (''PDBADMIN user not found in ' || name || ' PDB'');' || CHR(10)
+        || '    ELSE' || CHR(10)
+        || '      RAISE;' || CHR(10)
+        || '    END IF;' || CHR(10)
+        || 'END;' || CHR(10)
+        || '/'
+  FROM     v\$pdbs
+  WHERE    open_mode = 'READ WRITE'
+  ORDER BY name;
+EOF
+
+sed -i -e 's|no rows selected|PROMPT No open PDBs found|' "${set_pdbadmin_pw}"
+
+echo 'EXIT' | sqlplus -s / as sysdba @"${set_pdbadmin_pw}"
+
+echo 'Done setting PDBADMIN password in open PDBs'

--- a/OracleDatabase/12.2.0.1/scripts/setPassword.sh
+++ b/OracleDatabase/12.2.0.1/scripts/setPassword.sh
@@ -1,26 +1,54 @@
 #!/bin/bash
-# LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2016 Oracle and/or its affiliates. All rights reserved.
-# 
+# Copyright (c) 2023 Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at
+# https://oss.oracle.com/licenses/upl.
+#
 # Since: November, 2016
 # Author: gerald.venzl@oracle.com
-# Description: Sets the password for sys, system and pdb_admin
+# Description: Sets the password for sys, system and pdbadmin
 #
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
-# 
+#
+
+# Abort on any error
+set -Eeuo pipefail
 
 ORACLE_PWD=$1
-ORACLE_SID="`grep $ORACLE_HOME /etc/oratab | cut -d: -f1`"
-ORACLE_PDB="`ls -dl $ORACLE_BASE/oradata/$ORACLE_SID/*/ | grep -v pdbseed | awk '{print $9}' | cut -d/ -f6`"
-ORAENV_ASK=NO
-source oraenv
 
 sqlplus / as sysdba << EOF
-      ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
-      ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
-      ALTER SESSION SET CONTAINER=$ORACLE_PDB;
-      ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD";
-      exit;
+  ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
+  ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
 EOF
 
+echo 'Setting PDBADMIN password in open PDBs'
+
+set_pdbadmin_pw=$(mktemp)
+trap 'rm -f "${set_pdbadmin_pw}"' EXIT
+
+sqlplus -s / as sysdba > "${set_pdbadmin_pw}" << EOF
+  SET HEADING OFF LINESIZE 120 PAGESIZE 0
+  SELECT   'ALTER SESSION SET CONTAINER = ' || name || ';' || CHR(10)
+        || 'SET SERVEROUTPUT ON' || CHR(10)
+        || 'BEGIN' || CHR(10)
+        || '  EXECUTE IMMEDIATE ''ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD"'';' || CHR(10)
+        || '  DBMS_OUTPUT.PUT_LINE (''Set PDBADMIN password in ' || name || ' PDB'');' || CHR(10)
+        || 'EXCEPTION' || CHR(10)
+        || '  WHEN OTHERS THEN' || CHR(10)
+        || '    IF SQLCODE = -1918 THEN' || CHR(10)
+        || '      DBMS_OUTPUT.PUT_LINE (''PDBADMIN user not found in ' || name || ' PDB'');' || CHR(10)
+        || '    ELSE' || CHR(10)
+        || '      RAISE;' || CHR(10)
+        || '    END IF;' || CHR(10)
+        || 'END;' || CHR(10)
+        || '/'
+  FROM     v\$pdbs
+  WHERE    open_mode = 'READ WRITE'
+  ORDER BY name;
+EOF
+
+sed -i -e 's|no rows selected|PROMPT No open PDBs found|' "${set_pdbadmin_pw}"
+
+echo 'EXIT' | sqlplus -s / as sysdba @"${set_pdbadmin_pw}"
+
+echo 'Done setting PDBADMIN password in open PDBs'

--- a/OracleDatabase/18.3.0/scripts/setPassword.sh
+++ b/OracleDatabase/18.3.0/scripts/setPassword.sh
@@ -1,26 +1,54 @@
 #!/bin/bash
-# LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
-# 
+# Copyright (c) 2023 Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at
+# https://oss.oracle.com/licenses/upl.
+#
 # Since: November, 2016
 # Author: gerald.venzl@oracle.com
-# Description: Sets the password for sys, system and pdb_admin
+# Description: Sets the password for sys, system and pdbadmin
 #
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
-# 
+#
+
+# Abort on any error
+set -Eeuo pipefail
 
 ORACLE_PWD=$1
-ORACLE_SID="`grep $ORACLE_HOME /etc/oratab | cut -d: -f1`"
-ORACLE_PDB="`ls -dl $ORACLE_BASE/oradata/$ORACLE_SID/*/ | grep -v pdbseed | awk '{print $9}' | cut -d/ -f6`"
-ORAENV_ASK=NO
-source oraenv
 
 sqlplus / as sysdba << EOF
-      ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
-      ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
-      ALTER SESSION SET CONTAINER=$ORACLE_PDB;
-      ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD";
-      exit;
+  ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
+  ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
 EOF
 
+echo 'Setting PDBADMIN password in open PDBs'
+
+set_pdbadmin_pw=$(mktemp)
+trap 'rm -f "${set_pdbadmin_pw}"' EXIT
+
+sqlplus -s / as sysdba > "${set_pdbadmin_pw}" << EOF
+  SET HEADING OFF LINESIZE 120 PAGESIZE 0
+  SELECT   'ALTER SESSION SET CONTAINER = ' || name || ';' || CHR(10)
+        || 'SET SERVEROUTPUT ON' || CHR(10)
+        || 'BEGIN' || CHR(10)
+        || '  EXECUTE IMMEDIATE ''ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD"'';' || CHR(10)
+        || '  DBMS_OUTPUT.PUT_LINE (''Set PDBADMIN password in ' || name || ' PDB'');' || CHR(10)
+        || 'EXCEPTION' || CHR(10)
+        || '  WHEN OTHERS THEN' || CHR(10)
+        || '    IF SQLCODE = -1918 THEN' || CHR(10)
+        || '      DBMS_OUTPUT.PUT_LINE (''PDBADMIN user not found in ' || name || ' PDB'');' || CHR(10)
+        || '    ELSE' || CHR(10)
+        || '      RAISE;' || CHR(10)
+        || '    END IF;' || CHR(10)
+        || 'END;' || CHR(10)
+        || '/'
+  FROM     v\$pdbs
+  WHERE    open_mode = 'READ WRITE'
+  ORDER BY name;
+EOF
+
+sed -i -e 's|no rows selected|PROMPT No open PDBs found|' "${set_pdbadmin_pw}"
+
+echo 'EXIT' | sqlplus -s / as sysdba @"${set_pdbadmin_pw}"
+
+echo 'Done setting PDBADMIN password in open PDBs'

--- a/OracleDatabase/18.4.0-XE/scripts/setPassword.sh
+++ b/OracleDatabase/18.4.0-XE/scripts/setPassword.sh
@@ -1,26 +1,54 @@
 #!/bin/bash
-# LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
-# 
+# Copyright (c) 2023 Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at
+# https://oss.oracle.com/licenses/upl.
+#
 # Since: November, 2016
 # Author: gerald.venzl@oracle.com
-# Description: Sets the password for sys, system and pdb_admin
+# Description: Sets the password for sys, system and pdbadmin
 #
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
-# 
+#
+
+# Abort on any error
+set -Eeuo pipefail
 
 ORACLE_PWD=$1
-ORACLE_SID="`grep $ORACLE_HOME /etc/oratab | cut -d: -f1`"
-ORACLE_PDB="`ls -dl $ORACLE_BASE/oradata/$ORACLE_SID/*/ | grep -v pdbseed | awk '{print $9}' | cut -d/ -f6`"
-ORAENV_ASK=NO
-source oraenv
 
 sqlplus / as sysdba << EOF
-      ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
-      ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
-      ALTER SESSION SET CONTAINER=$ORACLE_PDB;
-      ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD";
-      exit;
+  ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
+  ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
 EOF
 
+echo 'Setting PDBADMIN password in open PDBs'
+
+set_pdbadmin_pw=$(mktemp)
+trap 'rm -f "${set_pdbadmin_pw}"' EXIT
+
+sqlplus -s / as sysdba > "${set_pdbadmin_pw}" << EOF
+  SET HEADING OFF LINESIZE 120 PAGESIZE 0
+  SELECT   'ALTER SESSION SET CONTAINER = ' || name || ';' || CHR(10)
+        || 'SET SERVEROUTPUT ON' || CHR(10)
+        || 'BEGIN' || CHR(10)
+        || '  EXECUTE IMMEDIATE ''ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD"'';' || CHR(10)
+        || '  DBMS_OUTPUT.PUT_LINE (''Set PDBADMIN password in ' || name || ' PDB'');' || CHR(10)
+        || 'EXCEPTION' || CHR(10)
+        || '  WHEN OTHERS THEN' || CHR(10)
+        || '    IF SQLCODE = -1918 THEN' || CHR(10)
+        || '      DBMS_OUTPUT.PUT_LINE (''PDBADMIN user not found in ' || name || ' PDB'');' || CHR(10)
+        || '    ELSE' || CHR(10)
+        || '      RAISE;' || CHR(10)
+        || '    END IF;' || CHR(10)
+        || 'END;' || CHR(10)
+        || '/'
+  FROM     v\$pdbs
+  WHERE    open_mode = 'READ WRITE'
+  ORDER BY name;
+EOF
+
+sed -i -e 's|no rows selected|PROMPT No open PDBs found|' "${set_pdbadmin_pw}"
+
+echo 'EXIT' | sqlplus -s / as sysdba @"${set_pdbadmin_pw}"
+
+echo 'Done setting PDBADMIN password in open PDBs'

--- a/OracleDatabase/19.3.0/scripts/setPassword.sh
+++ b/OracleDatabase/19.3.0/scripts/setPassword.sh
@@ -1,26 +1,54 @@
 #!/bin/bash
-# LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
-# 
+# Copyright (c) 2023 Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at
+# https://oss.oracle.com/licenses/upl.
+#
 # Since: November, 2016
 # Author: gerald.venzl@oracle.com
-# Description: Sets the password for sys, system and pdb_admin
+# Description: Sets the password for sys, system and pdbadmin
 #
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
-# 
+#
+
+# Abort on any error
+set -Eeuo pipefail
 
 ORACLE_PWD=$1
-ORACLE_SID="`grep $ORACLE_HOME /etc/oratab | cut -d: -f1`"
-ORACLE_PDB="`ls -dl $ORACLE_BASE/oradata/$ORACLE_SID/*/ | grep -v pdbseed | awk '{print $9}' | cut -d/ -f6`"
-ORAENV_ASK=NO
-source oraenv
 
 sqlplus / as sysdba << EOF
-      ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
-      ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
-      ALTER SESSION SET CONTAINER=$ORACLE_PDB;
-      ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD";
-      exit;
+  ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
+  ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
 EOF
 
+echo 'Setting PDBADMIN password in open PDBs'
+
+set_pdbadmin_pw=$(mktemp)
+trap 'rm -f "${set_pdbadmin_pw}"' EXIT
+
+sqlplus -s / as sysdba > "${set_pdbadmin_pw}" << EOF
+  SET HEADING OFF LINESIZE 120 PAGESIZE 0
+  SELECT   'ALTER SESSION SET CONTAINER = ' || name || ';' || CHR(10)
+        || 'SET SERVEROUTPUT ON' || CHR(10)
+        || 'BEGIN' || CHR(10)
+        || '  EXECUTE IMMEDIATE ''ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD"'';' || CHR(10)
+        || '  DBMS_OUTPUT.PUT_LINE (''Set PDBADMIN password in ' || name || ' PDB'');' || CHR(10)
+        || 'EXCEPTION' || CHR(10)
+        || '  WHEN OTHERS THEN' || CHR(10)
+        || '    IF SQLCODE = -1918 THEN' || CHR(10)
+        || '      DBMS_OUTPUT.PUT_LINE (''PDBADMIN user not found in ' || name || ' PDB'');' || CHR(10)
+        || '    ELSE' || CHR(10)
+        || '      RAISE;' || CHR(10)
+        || '    END IF;' || CHR(10)
+        || 'END;' || CHR(10)
+        || '/'
+  FROM     v\$pdbs
+  WHERE    open_mode = 'READ WRITE'
+  ORDER BY name;
+EOF
+
+sed -i -e 's|no rows selected|PROMPT No open PDBs found|' "${set_pdbadmin_pw}"
+
+echo 'EXIT' | sqlplus -s / as sysdba @"${set_pdbadmin_pw}"
+
+echo 'Done setting PDBADMIN password in open PDBs'

--- a/OracleDatabase/21.3.0-XE/scripts/setPassword.sh
+++ b/OracleDatabase/21.3.0-XE/scripts/setPassword.sh
@@ -1,26 +1,54 @@
 #!/bin/bash
-# LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
-# 
+# Copyright (c) 2023 Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at
+# https://oss.oracle.com/licenses/upl.
+#
 # Since: November, 2016
 # Author: gerald.venzl@oracle.com
-# Description: Sets the password for sys, system and pdb_admin
+# Description: Sets the password for sys, system and pdbadmin
 #
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
-# 
+#
+
+# Abort on any error
+set -Eeuo pipefail
 
 ORACLE_PWD=$1
-ORACLE_SID="`grep $ORACLE_HOME /etc/oratab | cut -d: -f1`"
-ORACLE_PDB="`ls -dl $ORACLE_BASE/oradata/$ORACLE_SID/*/ | grep -v pdbseed | awk '{print $9}' | cut -d/ -f6`"
-ORAENV_ASK=NO
-source oraenv
 
 sqlplus / as sysdba << EOF
-      ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
-      ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
-      ALTER SESSION SET CONTAINER=$ORACLE_PDB;
-      ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD";
-      exit;
+  ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
+  ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
 EOF
 
+echo 'Setting PDBADMIN password in open PDBs'
+
+set_pdbadmin_pw=$(mktemp)
+trap 'rm -f "${set_pdbadmin_pw}"' EXIT
+
+sqlplus -s / as sysdba > "${set_pdbadmin_pw}" << EOF
+  SET HEADING OFF LINESIZE 120 PAGESIZE 0
+  SELECT   'ALTER SESSION SET CONTAINER = ' || name || ';' || CHR(10)
+        || 'SET SERVEROUTPUT ON' || CHR(10)
+        || 'BEGIN' || CHR(10)
+        || '  EXECUTE IMMEDIATE ''ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD"'';' || CHR(10)
+        || '  DBMS_OUTPUT.PUT_LINE (''Set PDBADMIN password in ' || name || ' PDB'');' || CHR(10)
+        || 'EXCEPTION' || CHR(10)
+        || '  WHEN OTHERS THEN' || CHR(10)
+        || '    IF SQLCODE = -1918 THEN' || CHR(10)
+        || '      DBMS_OUTPUT.PUT_LINE (''PDBADMIN user not found in ' || name || ' PDB'');' || CHR(10)
+        || '    ELSE' || CHR(10)
+        || '      RAISE;' || CHR(10)
+        || '    END IF;' || CHR(10)
+        || 'END;' || CHR(10)
+        || '/'
+  FROM     v\$pdbs
+  WHERE    open_mode = 'READ WRITE'
+  ORDER BY name;
+EOF
+
+sed -i -e 's|no rows selected|PROMPT No open PDBs found|' "${set_pdbadmin_pw}"
+
+echo 'EXIT' | sqlplus -s / as sysdba @"${set_pdbadmin_pw}"
+
+echo 'Done setting PDBADMIN password in open PDBs'

--- a/OracleDatabase/21.3.0/scripts/setPassword.sh
+++ b/OracleDatabase/21.3.0/scripts/setPassword.sh
@@ -1,26 +1,54 @@
 #!/bin/bash
-# LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
-# 
+# Copyright (c) 2023 Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at
+# https://oss.oracle.com/licenses/upl.
+#
 # Since: November, 2016
 # Author: gerald.venzl@oracle.com
-# Description: Sets the password for sys, system and pdb_admin
+# Description: Sets the password for sys, system and pdbadmin
 #
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
-# 
+#
+
+# Abort on any error
+set -Eeuo pipefail
 
 ORACLE_PWD=$1
-ORACLE_SID="`grep $ORACLE_HOME /etc/oratab | cut -d: -f1`"
-ORACLE_PDB="`ls -dl $ORACLE_BASE/oradata/$ORACLE_SID/*/ | grep -v pdbseed | awk '{print $9}' | cut -d/ -f6`"
-ORAENV_ASK=NO
-source oraenv
 
 sqlplus / as sysdba << EOF
-      ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
-      ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
-      ALTER SESSION SET CONTAINER=$ORACLE_PDB;
-      ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD";
-      exit;
+  ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
+  ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
 EOF
 
+echo 'Setting PDBADMIN password in open PDBs'
+
+set_pdbadmin_pw=$(mktemp)
+trap 'rm -f "${set_pdbadmin_pw}"' EXIT
+
+sqlplus -s / as sysdba > "${set_pdbadmin_pw}" << EOF
+  SET HEADING OFF LINESIZE 120 PAGESIZE 0
+  SELECT   'ALTER SESSION SET CONTAINER = ' || name || ';' || CHR(10)
+        || 'SET SERVEROUTPUT ON' || CHR(10)
+        || 'BEGIN' || CHR(10)
+        || '  EXECUTE IMMEDIATE ''ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD"'';' || CHR(10)
+        || '  DBMS_OUTPUT.PUT_LINE (''Set PDBADMIN password in ' || name || ' PDB'');' || CHR(10)
+        || 'EXCEPTION' || CHR(10)
+        || '  WHEN OTHERS THEN' || CHR(10)
+        || '    IF SQLCODE = -1918 THEN' || CHR(10)
+        || '      DBMS_OUTPUT.PUT_LINE (''PDBADMIN user not found in ' || name || ' PDB'');' || CHR(10)
+        || '    ELSE' || CHR(10)
+        || '      RAISE;' || CHR(10)
+        || '    END IF;' || CHR(10)
+        || 'END;' || CHR(10)
+        || '/'
+  FROM     v\$pdbs
+  WHERE    open_mode = 'READ WRITE'
+  ORDER BY name;
+EOF
+
+sed -i -e 's|no rows selected|PROMPT No open PDBs found|' "${set_pdbadmin_pw}"
+
+echo 'EXIT' | sqlplus -s / as sysdba @"${set_pdbadmin_pw}"
+
+echo 'Done setting PDBADMIN password in open PDBs'

--- a/OracleDatabase/23ai-Free/scripts/setPassword.sh
+++ b/OracleDatabase/23ai-Free/scripts/setPassword.sh
@@ -11,17 +11,44 @@
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 #
 
+# Abort on any error
+set -Eeuo pipefail
+
 ORACLE_PWD=$1
-ORACLE_SID="`grep $ORACLE_HOME /etc/oratab | cut -d: -f1`"
-ORACLE_PDB="`ls -dl $ORACLE_BASE/oradata/$ORACLE_SID/*/ | grep -v pdbseed | awk '{print $9}' | cut -d/ -f6`"
-ORAENV_ASK=NO
-source oraenv
 
 sqlplus / as sysdba << EOF
-      ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
-      ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
-      ALTER SESSION SET CONTAINER=$ORACLE_PDB;
-      ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD";
-      exit;
+  ALTER USER SYS IDENTIFIED BY "$ORACLE_PWD";
+  ALTER USER SYSTEM IDENTIFIED BY "$ORACLE_PWD";
 EOF
 
+echo 'Setting PDBADMIN password in open PDBs'
+
+set_pdbadmin_pw=$(mktemp)
+trap 'rm -f "${set_pdbadmin_pw}"' EXIT
+
+sqlplus -s / as sysdba > "${set_pdbadmin_pw}" << EOF
+  SET HEADING OFF LINESIZE 120 PAGESIZE 0
+  SELECT   'ALTER SESSION SET CONTAINER = ' || name || ';' || CHR(10)
+        || 'SET SERVEROUTPUT ON' || CHR(10)
+        || 'BEGIN' || CHR(10)
+        || '  EXECUTE IMMEDIATE ''ALTER USER PDBADMIN IDENTIFIED BY "$ORACLE_PWD"'';' || CHR(10)
+        || '  DBMS_OUTPUT.PUT_LINE (''Set PDBADMIN password in ' || name || ' PDB'');' || CHR(10)
+        || 'EXCEPTION' || CHR(10)
+        || '  WHEN OTHERS THEN' || CHR(10)
+        || '    IF SQLCODE = -1918 THEN' || CHR(10)
+        || '      DBMS_OUTPUT.PUT_LINE (''PDBADMIN user not found in ' || name || ' PDB'');' || CHR(10)
+        || '    ELSE' || CHR(10)
+        || '      RAISE;' || CHR(10)
+        || '    END IF;' || CHR(10)
+        || 'END;' || CHR(10)
+        || '/'
+  FROM     v\$pdbs
+  WHERE    open_mode = 'READ WRITE'
+  ORDER BY name;
+EOF
+
+sed -i -e 's|no rows selected|PROMPT No open PDBs found|' "${set_pdbadmin_pw}"
+
+echo 'EXIT' | sqlplus -s / as sysdba @"${set_pdbadmin_pw}"
+
+echo 'Done setting PDBADMIN password in open PDBs'

